### PR TITLE
[stdlib] SR-1519: Implement SE-0032: Add Sequence.first(predicate:)

### PR DESCRIPTION
**Pull request by russbishop**
#### What's in this pull request?

Implementation of SE-0032, adds `first()` to `Sequence` protocol along with default implementation.

I modified `FindTest` to give each of the elements a separate `identity` so the `find` test can verify it found the first element and not just any matching element.
#### Resolved bug number: ([SR-1519](https://bugs.swift.org/browse/SR-1519))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: stdlib/public/core/Sequence.swift [hunk 1]
```diff
@@ -950,13 +962,32 @@ extension Sequence {
   ///
   /// - Parameter body: A closure that takes an element of the sequence as a
   ///   parameter.
+  @warn_unused_result
```
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: Looks like an unrelated edit... Even more, `@warn_unused_result` does nothing now.

File: stdlib/public/core/Sequence.swift [hunk 2]
```diff
@@ -564,6 +564,18 @@ public protocol Sequence {
...
+  /// predicate or nil if no such element is found.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element is a match.
+  /// - Returns: The first match or `nil` if there was no match.
+  @warn_unused_result
+  func first(
+    predicate: @noescape (Iterator.Element) throws -> Bool
+  ) rethrows -> Iterator.Element?
```
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: For this part, we need a "dispatch" test in `validation-test/stdlib/SequenceType.swift.gyb` to ensure that we have a requirement.

File: stdlib/private/StdlibCollectionUnittest/CheckSequenceType.swift [hunk 3]
```diff
@@ -127,7 +127,11 @@ public struct FindTest {
   ) {
     self.expected = expected
     self.element = MinimalEquatableValue(element)
-    self.sequence = sequence.map(MinimalEquatableValue.init)
+    var elementIndex = 0
+    self.sequence = sequence.map {
+    	defer { elementIndex += 1 }
```
Comment (gribozavr) [inline stdlib/private/StdlibCollectionUnittest/CheckSequenceType.swift]: It is strongly discouraged to write side-effectful closures for map().  I'd recommend `sequence.enumerated().map { ... }`.

File: stdlib/public/core/Sequence.swift [hunk 4]
```diff
@@ -950,13 +962,32 @@ extension Sequence {
...
+  /// predicate or nil if no such element is found.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element is a match.
+  /// - Returns: The first match or `nil` if there was no match.
+  public func first(
+    predicate: @noescape (Iterator.Element) throws -> Bool
+  ) rethrows -> Iterator.Element? {
+    for element in self {
```
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: Base it on `forEach` instead?  `forEach` might be cheaper (by a constant factor) for some kinds of collections with complex iteration patterns, for example, trees.

File: validation-test/stdlib/SequenceType.swift.gyb [hunk 5]
```diff
@@ -617,6 +617,26 @@ SequenceTypeTests.test("contains/Predicate") {
 }
 
 //===----------------------------------------------------------------------===//
+// first()
+//===----------------------------------------------------------------------===//
+
+SequenceTypeTests.test("first/Predicate") {
```
Comment (gribozavr) [inline validation-test/stdlib/SequenceType.swift.gyb]: Please add this test to `stdlib/private/StdlibCollectionUnittest/CheckSequenceType.swift` instead, since `first(where:)` is a customization point and different sequences and collections can implement it differently.

File: stdlib/public/core/Sequence.swift [hunk 6]
```diff
@@ -564,6 +564,18 @@ public protocol Sequence {
...
+  /// Returns the first element of the sequence that satisfies the given
+  /// predicate or nil if no such element is found.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element is a match.
+  /// - Returns: The first match or `nil` if there was no match.
+  @warn_unused_result
+  func first(
+    predicate: @noescape (Iterator.Element) throws -> Bool
```
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: The signature is `first(where:)` according to the proposal.

File: stdlib/public/core/Sequence.swift (hunk 1 above)
Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: Sorry, my fault. I also didn't realize `@discardable_result` landed already.

File: stdlib/public/core/Sequence.swift (hunk 4 above)
Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: `forEach` can't break out of iterating the entire sequence which IMHO would be surprising behavior, especially for non-restartable or non-terminating sequences.

(I actually think the `forEach` API could use a termination flag for early exit but that's another discussion.)

File: validation-test/stdlib/SequenceType.swift.gyb (hunk 5 above)
Comment (russbishop) [inline validation-test/stdlib/SequenceType.swift.gyb]: I'll move it. It still isn't quite clear to me exactly where the different kinds of tests belong. Is that stuff documented anywhere?

File: stdlib/public/core/Sequence.swift (hunk 4 above)
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: > forEach can't break out of iterating the entire sequence

We can throw :)  I'm not sure if the semantic analysis will accept that and not confuse it with the rethrow semantics, but it is worth a shot.

File: stdlib/private/StdlibCollectionUnittest/CheckSequenceType.swift (hunk 3 above)
Comment (russbishop) [inline stdlib/private/StdlibCollectionUnittest/CheckSequenceType.swift]: No problem, I'll change it.

File: validation-test/stdlib/SequenceType.swift.gyb (hunk 5 above)
Comment (gribozavr) [inline validation-test/stdlib/SequenceType.swift.gyb]: Sorry, it is not documented.  The general principle is that we want to put all algorithm tests into the reusable `StdlibCollectionUnittest` module, so that we can maximize our algorithm and collection coverage.

File: stdlib/public/core/Sequence.swift (hunk 6 above)
Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: Also my fault for not realizing the ability to use keywords as parameters change already landed.

File: stdlib/public/core/Sequence.swift (hunk 2 above)
Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: Ahhh I looked at this again and I think I see what the dispatch test is doing. I'll fix it.

File: stdlib/public/core/Sequence.swift (hunk 4 above)
Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: This didn't work at some point in the past but rethrow has gotten smarter since then. I tested that it doesn't interfere with throwing custom errors, nor does it break rethrows for a non-throwing closure. 

I wanted to get your opinion on this before I push a commit since it looks "interesting" at first glance 😁



```swift
enum _ForEachStopIteration: ErrorProtocol {
    case stopIteration
}

extension Sequence {
    /// Calls the given closure on each element in the sequence in the same order
    /// as a `for`-`in` loop.
    ///
    /// - Parameter body: A closure that takes an element of the sequence as a
    ///   parameter. Return `true` from this closure to stop iteration.
    ///   Returning `false` will continue iteration.
    func _forEach(_ body: @noescape (Iterator.Element) throws -> Bool) rethrows {
        do {
            try self.forEach {
                if try body($0) {
                    throw _ForEachStopIteration.stopIteration
                }
            }
        } catch is _ForEachStopIteration { }
    }

    /// <insert find description here>
    public func find(where predicate: (Iterator.Element) throws -> Bool) rethrows -> Iterator.Element? {
        var foundElement: Iterator.Element? = nil
        try self._forEach {
            if try predicate($0) {
                foundElement = $0
                return true
            }
            return false
        }
        return foundElement
    }
}
```

**Comment by russbishop** [comment]
@gribozavr Pushed the other changes, all tests pass locally. Just waiting to get your opinion on the throw to stop iteration code.

File: validation-test/stdlib/SequenceType.swift.gyb [hunk 7]
```diff
@@ -945,6 +945,16 @@ SequenceTypeTests.test("forEach/dispatch") {
 }
 
 //===----------------------------------------------------------------------===//
+// first()
+//===----------------------------------------------------------------------===//
+
+SequenceTypeTests.test("first/dispatch") {
+  let tester = SequenceLog.dispatchTester([OpaqueValue(1)])
+  tester.first { return $0.value == 1 }
```
Comment (gribozavr) [inline validation-test/stdlib/SequenceType.swift.gyb]: `return` is probably not needed here.

File: stdlib/public/core/Sequence.swift (hunk 4 above)
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: This LGTM, except I'd probably prefer to inline `_forEach` into its only use site.  `return true` and `return false` are less readable than just throwing and catching, I think.

Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: That's fine with me. I was considering putting in a proposal for a public version of `forEach` that allows stopping but I don't care for returning a `Bool` from the closure and an `inout` parameter seemed strange so maybe it just doesn't fit as a stdlib API.

**Comment by russbishop** [comment]
@gribozavr I think this is ready to go unless you can think of anything else

File: stdlib/public/core/Sequence.swift [hunk 8]
```diff
@@ -959,6 +970,34 @@ extension Sequence {
   }
 }
 
+private enum _StopIteration: ErrorProtocol {
```
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: Sorry, one more thing -- we can't use `private` in the library due to a compiler bug.  Please use `internal`.

Please add a space before the colon to conform to the library coding style.

File: stdlib/public/core/Sequence.swift [hunk 9]
```diff
@@ -959,6 +970,34 @@ extension Sequence {
...
+    case stop
```
Comment (gribozavr) [inline stdlib/public/core/Sequence.swift]: Indentation is two spaces.

**Comment by gribozavr** [comment]
@russbishop Thank you, LGTM modulo two style comments!

File: stdlib/public/core/Sequence.swift (hunk 9 above)
Comment (russbishop) [inline stdlib/public/core/Sequence.swift]: Having separate tab guidelines between projects is killer :D

**Comment by russbishop** [comment]
@gribozavr Done!

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
@russbishop Could you take a look at the build errors?

**Comment by russbishop** [comment]
@gribozavr Sorry about that, pushed a fix.

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
CI failure is unrelated, merging.
